File: skills/actor-pattern/zig-compendium/04_actor_hierarchy.zig
// Example 04: Actor Hierarchy — Coordinator with Sub-Actors
//
// A coordinator (Application) manages sub-actors representing distinct
// operational modes. Demonstrates:
//   - Coordinator that owns one sub-actor at a time
//   - Sub-actors as tagged unions (SetupMode, RunningMode, ErrorMode)
//   - Fire-and-forget SetState (avoids deadlock)
//   - Sub-actor triggering its own replacement
//   - Transition through setup -> running -> error states
//
// Build & run:
//   cd zig-compendium && zig build-exe 04_actor_hierarchy.zig && ./04_actor_hierarchy

const std = @import("std");
const Channel = @import("channel.zig").Channel;
const print = std.debug.print;

const allocator = std.heap.page_allocator;

// ============================================================================
// Sub-Actor interfaces via tagged union
// ============================================================================

const SubActor = union(enum) {
    setup: *SetupMode,
    running: *RunningMode,
    err: ErrorMode,
};

// ============================================================================
// ErrorMode — Terminal state (no thread, just stores the error message)
// ============================================================================

const ErrorMode = struct {
    message: []const u8,
};

// ============================================================================
// SetupMode — Simulates a setup wizard
// ============================================================================

const SetupCmd = union(enum) {
    get_step: *Channel(i32),
    advance: *Channel(bool), // true = ok, false = already done
};

const SetupMode = struct {
    cmds: Channel(SetupCmd),
    thread: std.Thread,
    app: *Application,

    fn start(app: *Application) *SetupMode {
        const self = allocator.create(SetupMode) catch @panic("alloc");
        self.* = .{
            .cmds = Channel(SetupCmd).init(allocator, 64),
            .thread = undefined,
            .app = app,
        };
        self.thread = std.Thread.spawn(.{}, SetupMode.run, .{self}) catch @panic("spawn");
        return self;
    }

    fn run(self: *SetupMode) void {
        var step: i32 = 1;
        const total_steps: i32 = 3;

        while (self.cmds.recv()) |cmd| {
            switch (cmd) {
                .get_step => |reply| {
                    reply.send(step);
                },
                .advance => |reply| {
                    if (step < total_steps) {
                        step += 1;
                        reply.send(true);
                    } else {
                        reply.send(true);
                        // Fire-and-forget: tell coordinator to transition.
                        self.app.setState(.{ .running = RunningMode.start(self.app, "config-from-step-3") });
                    }
                },
            }
        }
    }

    fn getStep(self: *SetupMode) i32 {
        var reply = Channel(i32).init(allocator, 1);
        defer reply.deinit();
        self.cmds.send(.{ .get_step = &reply });
        return reply.recv() orelse 0;
    }

    fn advance(self: *SetupMode) void {
        var reply = Channel(bool).init(allocator, 1);
        defer reply.deinit();
        self.cmds.send(.{ .advance = &reply });
        _ = reply.recv();
    }

    fn stop(self: *SetupMode) void {
        self.cmds.close();
        self.thread.join();
        self.cmds.deinit();
        allocator.destroy(self);
    }
};

// ============================================================================
// RunningMode — Normal operation
// ============================================================================

const RunningCmd = union(enum) {
    get_status: *Channel(Status),
};

const Status = struct {
    config: []const u8,
    uptime_ms: i64,
};

const RunningMode = struct {
    cmds: Channel(RunningCmd),
    thread: std.Thread,
    config: []const u8,

    fn start(app: *Application, config: []const u8) *RunningMode {
        _ = app;
        const self = allocator.create(RunningMode) catch @panic("alloc");
        self.* = .{
            .cmds = Channel(RunningCmd).init(allocator, 64),
            .thread = undefined,
            .config = config,
        };
        self.thread = std.Thread.spawn(.{}, RunningMode.run, .{self}) catch @panic("spawn");
        return self;
    }

    fn run(self: *RunningMode) void {
        const start_ns = std.time.nanoTimestamp();

        while (self.cmds.recv()) |cmd| {
            switch (cmd) {
                .get_status => |reply| {
                    const elapsed_ns = std.time.nanoTimestamp() - start_ns;
                    const elapsed_ms: i64 = @intCast(@divFloor(elapsed_ns, 1_000_000));
                    reply.send(.{
                        .config = self.config,
                        .uptime_ms = elapsed_ms,
                    });
                },
            }
        }
    }

    fn getStatus(self: *RunningMode) Status {
        var reply = Channel(Status).init(allocator, 1);
        defer reply.deinit();
        self.cmds.send(.{ .get_status = &reply });
        return reply.recv() orelse .{ .config = "unknown", .uptime_ms = 0 };
    }

    fn stop(self: *RunningMode) void {
        self.cmds.close();
        self.thread.join();
        self.cmds.deinit();
        allocator.destroy(self);
    }
};

// ============================================================================
// Coordinator (Application)
// ============================================================================

const AppCmd = union(enum) {
    get_state: *Channel(SubActor),
    set_state: SubActor, // fire-and-forget, no reply channel
};

const Application = struct {
    cmds: Channel(AppCmd),
    thread: std.Thread,

    /// Create the coordinator, then start it with a SetupMode sub-actor.
    /// This two-phase init solves the circular dependency: SetupMode needs
    /// the Application pointer, and Application needs the initial SubActor.
    fn startWithSetup() *Application {
        const self = allocator.create(Application) catch @panic("alloc");
        self.* = .{
            .cmds = Channel(AppCmd).init(allocator, 64),
            .thread = undefined,
        };
        const setup = SetupMode.start(self);
        self.thread = std.Thread.spawn(.{}, Application.run, .{ self, SubActor{ .setup = setup } }) catch @panic("spawn");
        return self;
    }

    fn run(self: *Application, initial: SubActor) void {
        var current = initial;

        while (self.cmds.recv()) |cmd| {
            switch (cmd) {
                .get_state => |reply| {
                    reply.send(current);
                },
                .set_state => |new_state| {
                    // Stop the old sub-actor.
                    stopSubActor(current);
                    current = new_state;
                },
            }
        }

        // Cleanup on shutdown.
        stopSubActor(current);
    }

    fn getState(self: *Application) SubActor {
        var reply = Channel(SubActor).init(allocator, 1);
        defer reply.deinit();
        self.cmds.send(.{ .get_state = &reply });
        return reply.recv() orelse .{ .err = .{ .message = "coordinator gone" } };
    }

    // Fire-and-forget: no reply channel, no blocking.
    // This prevents the deadlock where a sub-actor synchronously waits for
    // SetState while the coordinator waits for the sub-actor to Stop.
    fn setState(self: *Application, new_state: SubActor) void {
        self.cmds.send(.{ .set_state = new_state });
    }

    fn stop(self: *Application) void {
        self.cmds.close();
        self.thread.join();
        self.cmds.deinit();
        allocator.destroy(self);
    }

    fn stopSubActor(sub: SubActor) void {
        switch (sub) {
            .setup => |s| s.stop(),
            .running => |r| r.stop(),
            .err => {},
        }
    }
};

// ============================================================================
// Demo
// ============================================================================

pub fn main() !void {
    // Start in SetupMode.
    const app = Application.startWithSetup();
    defer app.stop();

    // Walk through the setup wizard.
    for (0..5) |_| {
        const state = app.getState();
        switch (state) {
            .setup => |s| {
                const step = s.getStep();
                print("[Setup] On step {d}\n", .{step});
                s.advance();
                // Give the coordinator a moment to process fire-and-forget.
                std.Thread.sleep(50_000_000);
            },
            .running => |r| {
                const status = r.getStatus();
                print("[Running] config={s}, uptime={d}ms\n", .{ status.config, status.uptime_ms });
            },
            .err => |e| {
                print("[Error] {s}\n", .{e.message});
            },
        }
    }

    // Demonstrate error state.
    print("\n--- Triggering error state ---\n", .{});
    app.setState(.{ .err = .{ .message = "everything is on fire" } });
    std.Thread.sleep(50_000_000);

    const state = app.getState();
    switch (state) {
        .err => |e| {
            print("[ErrorMode] Parked with: {s}\n", .{e.message});
        },
        else => {
            print("[Unexpected] Not in error state\n", .{});
        },
    }
}

File: skills/actor-pattern/rust-compendium/src/bin/04_actor_hierarchy.rs
//! Example 04: Actor Hierarchy -- Coordinator with Sub-Actors
//!
//! A coordinator (Application) manages sub-actors representing distinct
//! operational modes. Demonstrates:
//!   - SubActor trait (analogous to Go's Stoppable interface)
//!   - StateBuilder type alias for deferred actor construction
//!   - Coordinator that owns one sub-actor at a time
//!   - Fire-and-forget SetState (avoids the deadlock trap)
//!   - RecoverableError for graceful fallback chains
//!   - ErrorApp as terminal state
//!   - Sub-actor triggering its own replacement
//!   - Coordinator forwarding typed queries to the current sub-actor
//!
//! Run: cargo run --bin 04_actor_hierarchy

use std::sync::mpsc::{self, Sender};
use std::thread;
use std::time::{Duration, Instant};

// ============================================================================
// Core Abstractions
// ============================================================================

/// Common interface for all managed sub-actors.
trait SubActor: Send {
    /// Signal the sub-actor to stop and wait for its thread to finish.
    fn stop_and_join(&mut self);

    /// Describe the current state for display.
    fn describe(&self) -> StateDescription;

    /// Try to advance (only meaningful for SetupMode, others return None).
    fn try_advance(&self) -> Option<Result<(), String>> {
        None
    }
}

/// What the coordinator returns when asked about current state.
#[derive(Debug)]
enum StateDescription {
    Setup { step: i32 },
    Running { status: String, config: String },
    Error { msg: String },
}

/// Defers actor construction until the coordinator calls it.
type StateBuilder = Box<dyn FnOnce(&AppHandle) -> Result<Box<dyn SubActor>, BuildError> + Send>;

/// An error that optionally carries a fallback builder.
struct BuildError {
    msg: String,
    fallback: Option<StateBuilder>,
}

impl std::fmt::Display for BuildError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", self.msg)
    }
}

/// Try the builder, then one level of recovery, then ErrorApp.
fn build_state(app: &AppHandle, builder: StateBuilder) -> Box<dyn SubActor> {
    match builder(app) {
        Ok(actor) => actor,
        Err(BuildError {
            fallback: Some(next),
            msg,
        }) => {
            println!(
                "[Coordinator] Primary failed ({}), trying fallback...",
                msg
            );
            match next(app) {
                Ok(actor) => actor,
                Err(e) => Box::new(ErrorApp::new(e.msg)),
            }
        }
        Err(e) => Box::new(ErrorApp::new(e.msg)),
    }
}

// ============================================================================
// ErrorApp -- Terminal State
// ============================================================================

struct ErrorApp {
    err: String,
}

impl ErrorApp {
    fn new(err: String) -> Self {
        ErrorApp { err }
    }
}

impl SubActor for ErrorApp {
    fn stop_and_join(&mut self) {}

    fn describe(&self) -> StateDescription {
        StateDescription::Error {
            msg: self.err.clone(),
        }
    }
}

// ============================================================================
// Coordinator (Application)
// ============================================================================

enum AppCommand {
    /// Synchronous -- returns a description of the current state.
    GetState {
        reply: Sender<StateDescription>,
    },
    /// Synchronous -- advance the current sub-actor if it supports it.
    AdvanceSetup {
        reply: Sender<Option<Result<(), String>>>,
    },
    /// Fire-and-forget -- no reply channel. Critical to avoid deadlock
    /// when a sub-actor triggers its own replacement.
    SetState {
        builder: StateBuilder,
    },
}

/// Clone-able handle for interacting with the coordinator.
#[derive(Clone)]
struct AppHandle {
    tx: Sender<AppCommand>,
}

impl AppHandle {
    fn new(initial: StateBuilder) -> Self {
        let (tx, rx) = mpsc::channel::<AppCommand>();
        let handle = AppHandle { tx };
        let handle_for_thread = handle.clone();

        thread::spawn(move || {
            let mut current: Box<dyn SubActor> = build_state(&handle_for_thread, initial);

            while let Ok(cmd) = rx.recv() {
                match cmd {
                    AppCommand::GetState { reply } => {
                        let _ = reply.send(current.describe());
                    }
                    AppCommand::AdvanceSetup { reply } => {
                        let _ = reply.send(current.try_advance());
                    }
                    AppCommand::SetState { builder } => {
                        current.stop_and_join();
                        current = build_state(&handle_for_thread, builder);
                    }
                }
            }

            // Channel closed -- shut down current sub-actor.
            current.stop_and_join();
        });

        handle
    }

    fn get_state(&self) -> StateDescription {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx
            .send(AppCommand::GetState { reply: reply_tx })
            .unwrap();
        reply_rx.recv().unwrap()
    }

    fn advance_setup(&self) -> Option<Result<(), String>> {
        let (reply_tx, reply_rx) = mpsc::channel();
        self.tx
            .send(AppCommand::AdvanceSetup { reply: reply_tx })
            .unwrap();
        reply_rx.recv().unwrap()
    }

    /// Fire-and-forget. If the channel is closed, silently ignore.
    fn set_state(&self, builder: StateBuilder) {
        let _ = self.tx.send(AppCommand::SetState { builder });
    }
}

// ============================================================================
// Sub-Actor: SetupMode
// ============================================================================

enum SetupCommand {
    GetStep { reply: Sender<i32> },
    Advance { reply: Sender<Result<(), String>> },
}

struct SetupMode {
    tx: Option<Sender<SetupCommand>>,
    handle: Option<thread::JoinHandle<()>>,
}

impl SetupMode {
    fn new(app: &AppHandle) -> Result<Box<dyn SubActor>, BuildError> {
        let (tx, rx) = mpsc::channel::<SetupCommand>();
        let app_handle = app.clone();

        let join_handle = thread::spawn(move || {
            let mut step = 1;
            let total_steps = 3;

            while let Ok(cmd) = rx.recv() {
                match cmd {
                    SetupCommand::GetStep { reply } => {
                        let _ = reply.send(step);
                    }
                    SetupCommand::Advance { reply } => {
                        if step < total_steps {
                            step += 1;
                            let _ = reply.send(Ok(()));
                        } else {
                            let _ = reply.send(Ok(()));
                            // Trigger transition -- fire-and-forget.
                            let config = format!("config-from-step-{}", total_steps);
                            app_handle.set_state(Box::new(move |a: &AppHandle| {
                                RunningMode::new(a, &config)
                            }));
                        }
                    }
                }
            }
        });

        Ok(Box::new(SetupMode {
            tx: Some(tx),
            handle: Some(join_handle),
        }))
    }
}

impl SubActor for SetupMode {
    fn stop_and_join(&mut self) {
        // Drop the sender to signal the thread to exit.
        self.tx.take();
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
    }

    fn describe(&self) -> StateDescription {
        if let Some(ref tx) = self.tx {
            let (reply_tx, reply_rx) = mpsc::channel();
            if tx.send(SetupCommand::GetStep { reply: reply_tx }).is_ok() {
                if let Ok(step) = reply_rx.recv() {
                    return StateDescription::Setup { step };
                }
            }
        }
        StateDescription::Setup { step: 0 }
    }

    fn try_advance(&self) -> Option<Result<(), String>> {
        if let Some(ref tx) = self.tx {
            let (reply_tx, reply_rx) = mpsc::channel();
            if tx
                .send(SetupCommand::Advance { reply: reply_tx })
                .is_ok()
            {
                return Some(reply_rx.recv().unwrap_or(Err("recv failed".to_string())));
            }
        }
        Some(Err("setup actor stopped".to_string()))
    }
}

// ============================================================================
// Sub-Actor: RunningMode
// ============================================================================

enum RunningCommand {
    GetStatus { reply: Sender<String> },
}

struct RunningMode {
    tx: Option<Sender<RunningCommand>>,
    handle: Option<thread::JoinHandle<()>>,
    config: String, // immutable -- safe for direct reads
}

impl RunningMode {
    fn new(_app: &AppHandle, config: &str) -> Result<Box<dyn SubActor>, BuildError> {
        let (tx, rx) = mpsc::channel::<RunningCommand>();
        let cfg = config.to_string();
        let cfg_for_thread = cfg.clone();

        let join_handle = thread::spawn(move || {
            let up_since = Instant::now();

            while let Ok(cmd) = rx.recv() {
                match cmd {
                    RunningCommand::GetStatus { reply } => {
                        let uptime = up_since.elapsed();
                        let _ = reply.send(format!(
                            "running (config={}, uptime={:.0?})",
                            cfg_for_thread, uptime
                        ));
                    }
                }
            }
        });

        Ok(Box::new(RunningMode {
            tx: Some(tx),
            handle: Some(join_handle),
            config: cfg,
        }))
    }
}

impl SubActor for RunningMode {
    fn stop_and_join(&mut self) {
        self.tx.take();
        if let Some(h) = self.handle.take() {
            let _ = h.join();
        }
    }

    fn describe(&self) -> StateDescription {
        if let Some(ref tx) = self.tx {
            let (reply_tx, reply_rx) = mpsc::channel();
            if tx
                .send(RunningCommand::GetStatus { reply: reply_tx })
                .is_ok()
            {
                if let Ok(status) = reply_rx.recv() {
                    return StateDescription::Running {
                        status,
                        config: self.config.clone(),
                    };
                }
            }
        }
        StateDescription::Running {
            status: "stopped".to_string(),
            config: self.config.clone(),
        }
    }
}

// ============================================================================
// Demo
// ============================================================================

fn main() {
    // Start in SetupMode.
    let app = AppHandle::new(Box::new(|a: &AppHandle| SetupMode::new(a)));

    // Walk through the setup wizard.
    for _ in 0..4 {
        let desc = app.get_state();
        match &desc {
            StateDescription::Setup { step } => {
                println!("[Setup] On step {}", step);
                if let Some(result) = app.advance_setup() {
                    if let Err(e) = result {
                        println!("[Setup] Advance error: {}", e);
                    }
                }
                // Give fire-and-forget SetState a moment to process.
                thread::sleep(Duration::from_millis(50));
            }
            StateDescription::Running { status, config } => {
                println!("[Running] {}", status);
                println!("[Running] Config (direct getter): {}", config);
            }
            StateDescription::Error { msg } => {
                println!("[Error] {}", msg);
            }
        }
    }

    // Demonstrate RecoverableError: force a transition to a builder that fails
    // with recovery.
    println!("\n--- Triggering recoverable error ---");
    app.set_state(Box::new(|_a: &AppHandle| {
        Err(BuildError {
            msg: "primary database unavailable".to_string(),
            fallback: Some(Box::new(|a: &AppHandle| {
                println!("[Recovery] Falling back to read-only mode");
                RunningMode::new(a, "read-only-fallback")
            })),
        })
    }));
    thread::sleep(Duration::from_millis(50));

    match app.get_state() {
        StateDescription::Running { status, .. } => {
            println!("[After Recovery] {}", status);
        }
        other => println!("[After Recovery] unexpected: {:?}", other),
    }

    // Demonstrate unrecoverable error -> ErrorApp.
    println!("\n--- Triggering unrecoverable error ---");
    app.set_state(Box::new(|_a: &AppHandle| {
        Err(BuildError {
            msg: "everything is on fire".to_string(),
            fallback: Some(Box::new(|_a: &AppHandle| {
                Err(BuildError {
                    msg: "recovery also failed".to_string(),
                    fallback: None,
                })
            })),
        })
    }));
    thread::sleep(Duration::from_millis(50));

    match app.get_state() {
        StateDescription::Error { msg } => {
            println!("[ErrorApp] Parked with: {}", msg);
        }
        other => println!("[ErrorApp] unexpected: {:?}", other),
    }

    // Shutdown: drop the handle.
    drop(app);
    thread::sleep(Duration::from_millis(50));
}
